presto-kafka/src/main/java/io/prestosql/plugin/kafka/confluent/AvroConfluentRowDecoder.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.plugin.kafka.confluent;
+
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.google.common.collect.ImmutableMap;
+import io.confluent.kafka.schemaregistry.client.SchemaRegistryClient;
+import io.confluent.kafka.schemaregistry.client.rest.exceptions.RestClientException;
+import io.prestosql.decoder.DecoderColumnHandle;
+import io.prestosql.decoder.FieldValueProvider;
+import io.prestosql.decoder.RowDecoder;
+import io.prestosql.decoder.avro.AvroColumnDecoder;
+import io.prestosql.spi.PrestoException;
+import org.apache.avro.Schema;
+import org.apache.avro.generic.GenericDatumReader;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.avro.io.BinaryDecoder;
+import org.apache.avro.io.DecoderFactory;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.ByteBuffer;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import static com.google.common.base.Functions.identity;
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static io.prestosql.spi.StandardErrorCode.GENERIC_INTERNAL_ERROR;
+import static java.util.Objects.requireNonNull;
+
+public class AvroConfluentRowDecoder
+        implements RowDecoder
+{
+    public static final String NAME = "avro-confluent";
+    public static final int HEADER_LENGTH = 1 + Integer.BYTES;
+
+    private static ThreadLocal<BinaryDecoder> reuseDecoder = ThreadLocal.withInitial(() -> null);
+
+    private final Schema targetSchema;
+    private final Map<DecoderColumnHandle, AvroColumnDecoder> columnDecoders;
+    private final SchemaRegistryClient schemaRegistryClient;
+    private final LoadingCache<Integer, GenericDatumReader<?>> avroRecordReaderCache;
+
+    public AvroConfluentRowDecoder(Schema targetSchema, Set<DecoderColumnHandle> columns, SchemaRegistryClient schemaRegistryClient)
```

Comment:
What if we could have a custom supplier of `AvroRecordReader` to `AvroRowDecoder` , then we can re-use the code ?

Reply:
There are a few differences:
* The AvroRowDecoder expects a reader for GenericRecord, the confluent can either read GenericRecords or single java types, i.e. for a key with a simple schema like string or long (will use the subject as the column name).